### DISCUSSION TITLE
Round Two: Text Extraction in 2024

### DISCUSSION THREAD

**Discussion by oepen** [hplt-project/HPLT-textpipes#2] [General]
In early February, we are ready to launch our second end-to-end iteration, i.e. run our revised data processing (aka text extraction) pipeline on lots more data, train new sets of language and translation models, and evaluate these appropriately.  This discussion is intended to decide on the technical setup and schedule for the data processing part.

We are now looking at two common stages, at times called `warc2html`and `html2text`, which prepare the raw text collections for either monolingual or bilingual downstream processing.  These processes constitute the third stage, called _monotexting_ and _bitexting_ in HPLT.  Stage #1 will run close to the storage, i.e. on the dedicated NIRD and CESNET nodes that are available to the project.  For this stage, we still need to decide the _batch size_, i.e. how many WARC files each `warc2html` is to process.  For each such process, there will be a pair of (line-aligned) resulting output files, one of the HTML, another one for all the metadata.  The sum of these files will be much smaller than the original WARC collections and will form the input to Stage #2, which we expect to run on LUMI.

The _batch size_ should be chosen to make it practical and efficient to perform Stage #2.  We have yet to build the Stage #2 tool, but we assume (i) it will be in Python, (ii) perform some URL-based filtering, (iii) call out to Trafilatura for boilerplate removal, (iv) invoke FasterText for language identification, (v) perform sharding to support bitexting, and (vi) write out a suitable number of files in a suitable format (probably bifurcated by document language).  Based on contrastive benchmarking by Nikolay, we expect Stage #2 to be 20–30 more computationally expensive than the original `warc2text`.  In other words, we are making a substantive investment in terms of CPU hours here. 

In terms of storage, we expect that our roughly 4,5 pib of WARC files will require at least some 500 tib of storage after Stage #1.

 A rough calculation from a joint bus ride between Skeikampen and Oslo estimates the cost of Stage #1 to around 350,000 CPU hours, and up to some 10,000,000 CPU hours for Stage #2.  The latter is embarrassingly parallel.  To run it effectively on LUMI, we might want to break it into jobs that on average correspond to one single-node CPU day, which is only about 1600 jobs.  If each job processes one input file and parallelizes internally (e.g. using 256 Posix threads), it might seem sufficient to aim for Stage #1 to produce just a few thousand (pairs of) raw HTML files.  What would be the average estimated file size on this view?

And what about the output from Stage #2.  Ideally, we bifurcate by language (which may multiply our file count by about 200) at this point and shard in a manner suitable for bi-texting (which, in turn, will reduce the overall number of files).  How does sharding work internally, and can Stage #2 perform it directly, while going from raw HTML to raw text?

**Comment by akutuzov** [comment]
### Minutes for Oslo meeting (February 8, 2024)

- Do we want to store formatting in our data (markdown/HTML)?
- May be for the future
- What we want: segment boundaries and may be lists (with Trafilatura)

- The total size of the extracted HTML data infeasible on LUMI (up to a Petabyte)
- Thus, we will have to stream "macro-batches" from NIRD/CESNET to LUMI, process them, and delete the  input data from LUMI
- Also lets us avoid the file number limitations
- Deduplication right after Stage 2.
- @nvanva  as the main architect

**Comment by nvanva** [comment]
Summarizing the technical meeting on Stage 2 today. We now have the implementation of 2 steps from stage 2, langid and text extraction with boilerplate removal. Both run in parallel across several LUMI nodes, and across input lines on each node. We have agreed on the next TODOs:
1) Implement HTML matching with regexps removed from warc2text.  (@nvanva )
2) Develop staging logic. (@nvanva )
3) Develop giawarc-ing step. (@ZJaume )
4) Getting access to Norwegian LUMI and collecting statistics that might be interesting for the processed 1% sample (@rggdmonk)
5) Continue developing evaluation for boilerplate removal + langid. (@nvanva and @rggdmonk )

The slides are here (work in progress, will be updated, the link allows to leave comments there):
https://docs.google.com/presentation/d/1-AUBkFBwuPXUPMY756FyXhmFHr3sv3vm-dFJ7qG9Yq4/edit?usp=sharing

**Comment by nvanva** [comment]
Some principle decisions to make:
1) Which micro/macro-languages we select for iteration 2, how they correspond to language codes from open-lid201 (probably, we would like to merge similar languages, e.g. many Arabic languages that open-lid201 tries to distinguish)? After creating a list of languages and the mapping to open-lid201 classes we can work on the test set for (a subset of) those languages.
2) For Trafilatura we should select the extraction settings, for the pilot run we used: include_comments": False, "include_tables": False, "no_fallback": True
3) We should agree on the criterion for retrieving texts in language X. Previously all documents with the highest predicted probability for class X, given this probability is >0.5 were selected. Maybe we can think of some more theoretically or empirically sound criterion.

**Reply by akutuzov** [reply]
How many languages does ` open-lid201` distinguish?

**Reply by ZJaume** [reply]
@akutuzov 201

1. It was at sentence level, but I used lid201 a couple of weeks ago with Standard Arabic text, and the identifier predicted many other variants of Arabic, so it is probably messing up with all the variants because they are probably difficult to distinguish. We may also found this, like with previous langids (cld2, old fasttext), with Serbo-Croatian variants and Malay-Indonesian. Probably an option to avoid this, is to group all of this individual language codes into the inclusive codes. I think this is not something vert harmful from the perspective of people wanting to train LLMs for those macrolanguages, as they will probably train with all the variants together. One example is https://huggingface.co/classla/bcms-bertic. Another case maybe, is people looking for data in different variants of the same macrolanguage. I think in this case doesn't harm much, because people doing this kind of tasks usually has better tools to distinguishing between variants than our language identifier.

2. If stage2 is way faster than we predicted, maybe we can afford using slow version (a.k.a fallback)?

3. Yesterday I agreed on filtering by langid confidence at the a later step (maybe cleaning). But today I've been taking a look at Spanish and Catalan data and everything under 0.5 seems complete garbage. Also, I found in the 1% sample of wide10 that these under 0.5 docs are 4.4M out of 17M. So, I really think we could save space and computation if we early discard this, as we've been doing in previous projects.

I've also opened a couple of issues about the code:
https://github.com/hplt-project/warc2text-runner/issues/2
https://github.com/hplt-project/warc2text-runner/issues/1

**Reply by nvanva** [reply]
1) So we should think of a principled way to combine languages that both takes misclassification errors into account and aligns with the languages we have models for in Monotextor. Any ideas how to approach this problem?
2) In terms of wall clock time we can, but I'm not sure we will have enough core-hours. Rough estimates are 2.5M core-hours to process all data with the current settings, and @akutuzov says we will have only 3M core-hours for the next 6 months.
3) Looks like 0.5 is selected quit arbitrarily, are we sure this will work well for all languages? Does saving 25% of storage space worth it?

**Reply by akutuzov** [reply]
@nvanva we can ask for more CPU hours, but we need a compelling reason for this.

**Comment by nvanva** [comment]
Returning to our meeting today, two samples of HTMLs and texts, for all CC crawls and all wide crawls, are commited to https://github.com/hplt-project/warc2text-runner/tree/main/two/sample100/CC and https://github.com/hplt-project/warc2text-runner/tree/main/two/sample100/wide. All data for these samples is in https://github.com/hplt-project/warc2text-runner/tree/main/two/sample100/*.zst.

Also I've checked a couple of questions raised.

1) ~99% are decoded without UnicodeDecodeErrors:
![image](https://github.com/hplt-project/HPLT-WP2/assets/9987540/df7f8d50-0e10-4d2e-97e3-1881d9bc1be4)

2) Looking at one WARC and data extracted by warc2html, we now don't take HTTP response code into account and also don't dump them as metadata. This results in quit many placeholder pages that are now hard to filter out (7 out of 50 pages  I've recently annotated). At least for the 3rd iteration, we should try to modify warc2html to filter out such pages or dump response code for future filtering.

**Reply by ZJaume** [reply]
Would it be useful to know, maybe, the average number of encoding errors per document? If, the majority of documents with errors have only a few characters with errors, I think it makes sense to keep them decoding with `errors="ignore"`. If not, maybe just discard documents with errors.

**Reply by oepen** [reply]
regarding HTTP status codes, i would suggest to file this as a separate issue on warc2text.  naively, it seems tempting to discard WARC records for status codes other than 200 (ok) and maybe 203 or 206?

**Reply by ZJaume** [reply]
Yes, it is tempting. I've opened an issue https://github.com/bitextor/warc2text/issues/56

**Comment by nvanva** [comment]
Digging into the example mentioned by @ZJaume (https://github.com/hplt-project/warc2text-runner/blob/main/two/sample100/CC/27.txt), I found that by default Trafilatura falls back to some simple baseline method that leaves lots of boilerplate in case the main method produces too short text. So basically for this example we see the text extracted by the baseline, and the main method extracts nothing. My intuition is that we don't have much really valuable running text on this page, so extracting nothing may be better for this page rather than leaving all the boilerplate (giving a bit lower recall but much higher precision when considering the task of extracting "running text in natural language").

More details on the slides, starting from slide 11: https://docs.google.com/presentation/d/1-AUBkFBwuPXUPMY756FyXhmFHr3sv3vm-dFJ7qG9Yq4/edit?usp=sharing

Thus, I played with Trafilatura settings and for sample100 commited 2 new versions of extracted texts, using Trafilatura without fall back to the baseline (*.MIN_EXTRACTED_SIZE_0.txt) and using Trafilatura without falling back to the baseline but employing readability-lxml and justext in addition to the main extraction algorithm (*.no_fallback_False.txt). The last setting results in 2x slower processing though.

Comparing basic statistics for 4 methods (warc2text and 3 versions of Trafilatura):
https://docs.google.com/spreadsheets/d/1IHpQHWAPY2sAVnUPnKzp9IHmM4IqqDgIz0zcYWMI_4c/edit?usp=sharing

Without falling back to the baseline, Trafilatura returns empty texts for 20-40% of webpages. But looking at those pages it seems to me much more reasonable result than the baseline or warc2text gives if we want to extract running texts, not everything. Also it gives ~6-7x fewer lines, and ~3-4 fewer bytes, but ~2x longer lines on average. So looks like it removes lots of short lines that constitute a significant part of the corpus and probably contain lots of boilerplate. 

I see 2 reasonable ways to proceed:
1) We all take a look at 3 versions of text extracted for each page in sample100, and select the most reasonable settings based on this eyeballing.
2) We label (all or some of) the webpages from sample100, calculate metrics for each of the settings (can also experiment with more of them if evaluation is automated) and make the decision based on the metrics.

**Reply by ZJaume** [reply]
Just a note before I finish eyeballing the four approaches. The numbers of bytes/words per line are a bit biased in all trafilatura outputs if we don't address the `a`/`strong`/`span` tags introducing newlines problem.

**Reply by nvanva** [reply]
You mean biased compare to Warc2text? Or you imagine some true distribution?

Could you please refer me to a few files among *.MIN_EXTRACTED_SIZE_0.txt with this newlines problem? I will debug the main method on those files more.

**Reply by ZJaume** [reply]
I mean biased compared to both warc2text and some ideal distribution where we fix this issue.

Some documents that I found to be affected by this: wide/21, wide/50, CC/63. Besides these tags previsouly mentioned, also h1/2/3/4 inside tables results in a dash+newline+text.

EDIT: also CC/40

**Reply by ZJaume** [reply]
After taking a more deep look (not all samples), for me it seems that the best is `no_falback_False`. So the question would be if we can afford that in terms of computation resources?

If we have to choose between min_size=0 and default, I would say default. From all the documents that result in empty because of that min_size parameter (77 docs out of 200), there are 5 that won't be discarded by our current monotextor filters of `avg_num_words<5` and `doc_lenght<500`: wide/6.txt CC/3.txt CC/21.txt CC/45.txt CC/99.txt. Looking at this 5, all seem nice documents. So, I would prefer to use the default method and let our downstream filters in monotextor do their job, rather than loosing all of these 77 because they are empty.

**Reply by nvanva** [reply]
I can run this presumably best config on 1% to get a better estimate of the resources required.

Regarding those 77 documents, I'm not that convinced we want to keep them. We decided to employ trafilatura to remove as much boilerplate as possible at this stage, before langid, not to get the result similar to ones from warc2text. So I would prefer to sacrifice those 5 documents out of 200, rather than have only boilerplate extracted from the rest 72, plus extra boilerplate for some of the rest 200-77 documents.

**Reply by oepen** [reply]
i am away from the office today and tomorrow, and will be relatively busy the rest of the week.  i feel we should seek to launch Stage 2 processing as soon as possible.  sensible segmentation feels like a prerequisite.  no matter how well we end up balancing the anti-boilerplate mechanisms, i suspect we need to solve the problem of in-line elements like ```<strong>```or  ```<a>``` introducing spurious segment boundaries.   as i recall our previous analysis, that will require parameterizing and possibly modifying the code in ```xml.py```.  @nvanva, did you have a chance to look into this?

**Reply by nvanva** [reply]
What do you think about the settings I've described above? Oversegmentation happens mostly in the boilerplate, with these new setting the boilerplate is removed for short pages as well, and the outputs now look good to me (see files in  ending .MIN_EXTRACTED_SIZE_0.txt for some examples: https://github.com/hplt-project/warc2text-runner/tree/main/two/sample100/CC).

**Reply by ZJaume** [reply]
@nvanva debugged the trafilatura code and we know now that inline elements are not introducing newlines during the conversion of the "cleaned html" to text. However, there are some cases where it might seem that it is introducing extra newlines. From the above mentioned examples, CC/63 and wide/21 have extra newlines after dashes because the original html has an `h2` inside a row of a table. Then the wide/50, during html cleaning of boilerplate
```
Hey There @<a href="..." target="_blank">Visser</a> !!<br />
```
becomes
```
<p>Hey There@</p><p>Visser</p><p> !!</p> 
```
for some reason.

My concern was about trafilatura converting all those inline elements to newline during html to text conversion. But we checked and this doesn't seem to be the case.

After talking in today meeting, we are satisfied with the current outputs. @nvanva will estimate the cost for `no_fallback_False`, which seems the best output.

**Comment by nvanva** [comment]
Following our today discussion with @oepen, @akutuzov  and @davda54, this is how one can compare a sample of webpages processed with different Trafilatura settings. 

After cloning https://github.com/hplt-project/warc2text-runner/tree/main:
```
cd warc2text-runner/two/sample100
export C=CC X=33; open $C/$X.html; vimdiff $C/$X.MIN_EXTRACTED_SIZE_0.txt $C/$X.no_fallback_False.txt 
```
Choose C among CC and wide to switch between crawls, and change X to switch between sampled web pages. The selected html page will be opened in a web browser, and 2 versions of text extracted from this page with different settings will be shown in the terminal.

To switch between tabs in vimdiff: **ctrl+w <right/left arrow>**
To exit: **:qa**

**Reply by nvanva** [reply]
Upon the request from @spyysalo, new scripts for more convenient comparison are now available. You will be suggested to compare only different extraction results. Also after closing the current pair of files (:qa) next one is opened automatically. 
```
cd warc2text-runner/two/sample100
./compare.sh wide
./compare.sh CC
```

**Reply by nvanva** [reply]
Following @oepen question and recent discussion of FineWeb, you are welcome to compare our previously selected settings with the settings from FineWeb and share your impressions!
```
cd warc2text-runner/two/sample100
./compare.sh wide  fineweblike.txt  no_fallback_False.txt 
./compare.sh CC  fineweblike.txt  no_fallback_False.txt 
```

**Reply by ZJaume** [reply]
Looking at 40 random pages from both collections, It seems that in the pages where FineWeb configuration is able to extract more text, it is mainly boilerplate. Also the stats of bytes per line and words per line are lower:
```bash
$ cat {CC,wide}/*.no_fallback_False.txt | env LC_ALL=C awk '{a+=length($0); b+=split($0,x," ")}END{print a/NR " " b/NR}'
76.5389 8.86037
$ cat {CC,wide}/*.fineweblike.txt | env LC_ALL=C awk '{a+=length($0); b+=split($0,x," ")}END{print a/NR " " b/NR}'
36.631 4.21514
$ cat {CC,wide}/*.MIN_EXTRACTED_SIZE_0.txt | env LC_ALL=C awk '{a+=length($0); b+=split($0,x," ")}END{print a/NR " " b/NR}'
68.4108 8.83204
$ cat {,wide}/*.txt | env LC_ALL=C awk '{a+=length($0); b+=split($0,x," ")}END{print a/NR " " b/NR}'
39.1179 4.17713
$ cat {,wide}/*.warc2text | env LC_ALL=C awk '{a+=length($0); b+=split($0,x," ")}END{print a/NR " " b/NR}'
36.0209 3.7667
```

**Reply by nvanva** [reply]
Thanks a lot for inspecting! A looked at a few examples and also had an impression that FineWeb leaves much more boilerplate. Probably this is because they don't change the MIN_EXTRACTED_SIZE parameter in the config, and this results in the baseline method extracting texts for many documents that don't main content that is long enough.

**Reply by nvanva** [reply]
Actually the baseline is called for ~35% of all documents. 

To summarize their differences from our previously selected configuration:
1) they leave the default vale MIN_EXTRACTED_SIZE=250, we set it to 0; thus, for webpages where the main algorithms couldn't extract more than 250 characters of the main content, they extract (mostly boilerplate) with the baseline method
2) they set favour_precision=True, we leave this False as in the default;
3) they leave the default include_tables=True, we set it to False; this is based on my intuition that we probably don't want tables, but maybe I'm wrong?

**Reply by nvanva** [reply]
And the best thing is that trafilatura 1.8.0 and 1.8.1 with exactly the same settings give different results for ~20% of pages :)
![image](https://github.com/hplt-project/HPLT-WP2/assets/9987540/2b44645e-00fe-4301-a26e-8f18e83928b4)

**Reply by nvanva** [reply]
A potentially interesting flag which Fineweb enables and we do not is favour_precision. 

Following our discussion today, we decided to leave more manual comparisons of different hyperparameters for the future for the sake of time. But in case anyone is interested, this is how you can compare two configurations differing only by the value of favour_precision.
```
cd warc2text-runner/two/sample100
./compare.sh CC no_fallback_False.txt search_fineweblike_ms0_notables_traf1.8.0.txt
./compare.sh wide no_fallback_False.txt search_fineweblike_ms0_notables_traf1.8.0.txt
```
